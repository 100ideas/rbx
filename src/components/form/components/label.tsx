import { cx } from "emotion";
import React from "react";

import modifiers, { ModifierProps } from "modifiers";

export type LabelModifierProps = Partial<{
  children: React.ReactNode;
  className: string;
  htmlFor: string;
  size: "small" | "medium" | "large";
  style: {};
}>;

export type LabelProps = ModifierProps &
  LabelModifierProps &
  Partial<Omit<React.ComponentPropsWithoutRef<"label">, "unselectable">>;

const Label = React.forwardRef<HTMLLabelElement, LabelProps>(
  ({ children, className, size, ...allProps }, ref) => {
    const props = modifiers.clean(allProps);
    return (
      <label
        {...props}
        ref={ref}
        className={cx("label", modifiers.classNames(allProps), className, {
          [`is-${size}`]: size,
        })}
      >
        {children}
      </label>
    );
  },
);
Label.defaultProps = {
  children: null,
};

export default Label;
