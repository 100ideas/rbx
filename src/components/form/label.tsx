import { cx } from "emotion";
import React from "react";

import { ModifierProps, modify } from "@/modifiers";

export type LabelModifierProps = Partial<{
  htmlFor: string;
  size: "small" | "medium" | "large";
}>;

export type LabelProps = Prefer<
  ModifierProps & LabelModifierProps,
  React.HTMLAttributes<HTMLLabelElement>
>;

export const Label = React.forwardRef<HTMLLabelElement, LabelProps>(
  (props, ref) => {
    const { size, ...rest } = modify(props);
    rest.className = cx("label", rest.className, {
      [`is-${size}`]: size,
    });
    return <label {...rest} ref={ref} />;
  },
);
