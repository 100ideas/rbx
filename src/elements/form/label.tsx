import classNames from "classNames";
import * as PropTypes from "prop-types";
import * as React from "react";

import { forwardRefAs, Generic, HelpersProps } from "src/base";
import { tuple } from "src/utils";
import { Checkbox } from "./checkbox";
import { Radio } from "./radio";

export const LABEL_SIZES = tuple("small", "medium", "large");
export type LabelSizes = (typeof LABEL_SIZES)[number];

export type LabelModifierProps = Partial<{
  disabled: boolean;
  size: LabelSizes;
}>;

export type LabelProps = HelpersProps & LabelModifierProps;

const propTypes = {
  disabled: PropTypes.bool,
  size: PropTypes.oneOf(LABEL_SIZES),
};

const identifyLabelDiscriminator = (children: React.ReactNode) => {
  let discriminator = "label";

  React.Children.forEach(children, (child, i) => {
    if (typeof child === "object") {
      if (
        child.type === Checkbox ||
        (child.type === "input" &&
          (child.props as React.InputHTMLAttributes<Element>).type ===
            "checkbox")
      ) {
        discriminator = "checkbox";
      } else if (
        child.type === Radio ||
        (child.type === "input" &&
          (child.props as React.InputHTMLAttributes<Element>).type === "radio")
      ) {
        discriminator = "radio";
      } else if (child.type === React.Fragment) {
        discriminator = identifyLabelDiscriminator(
          (child.props as React.ComponentPropsWithoutRef<typeof React.Fragment>)
            .children,
        );
      }
    }
  });

  return discriminator;
};

export const Label = Object.assign(
  forwardRefAs<LabelProps, "label">(
    ({ className, disabled, size, ...rest }, ref) => {
      const discriminator = identifyLabelDiscriminator(rest.children);

      return (
        <Generic
          className={classNames(
            {
              [`${discriminator}`]: discriminator,
              "is-disabled": disabled,
              [`is-${size}`]: size,
            },
            className,
          )}
          ref={ref}
          {...rest}
        />
      );
    },
    { as: "label" },
  ),
  { propTypes },
);
