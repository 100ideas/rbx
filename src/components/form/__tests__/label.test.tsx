import React from "react";
import renderer from "react-test-renderer";

import { Label } from "../label";

describe("Label component", () => {
  it("should exist", () => {
    expect(Label).toMatchSnapshot();
  });

  it("should have label classname", () => {
    const component = renderer.create(
      <Label>
        Test <a>Give me</a>
      </Label>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat classname in props with Bulma classname", () => {
    const component = renderer.create(
      <Label className="other-class this-is-a-test">
        <p>Default</p>
      </Label>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should use inline styles", () => {
    const component = renderer.create(
      <Label style={{ height: 250 }}>
        <p>Default</p>
      </Label>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
