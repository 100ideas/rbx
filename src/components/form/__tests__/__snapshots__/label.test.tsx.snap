// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label component should concat classname in props with Bulma classname 1`] = `
<label
  className="label other-class this-is-a-test"
>
  <p>
    Default
  </p>
</label>
`;

exports[`Label component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[`Label component should have label classname 1`] = `
<label
  className="label"
>
  Test 
  <a>
    Give me
  </a>
</label>
`;

exports[`Label component should use inline styles 1`] = `
<label
  className="label"
  style={
    Object {
      "height": 250,
    }
  }
>
  <p>
    Default
  </p>
</label>
`;
